Build.ps1

```diff
@@ -32,149 +32,145 @@ param(
+$catchAllTags = [System.Linq.Enumerable]::Except([string[]]$availableTags, [string[]]$knownTags)
+$catchAllTags | ForEach-Object { $tags.Add($_) > $null }
+
+foreach ($wv in $WindowsVersion) 
 {
-    $tags.Add("mssql-developer:2017-windowsservercore-$($wv)") > $null
-    $tags.Add("sitecore-certificates:latest-nanoserver-$($windowsVersionMapping[$wv])") > $null
-    $tags.Add("sitecore-openjdk:8-nanoserver-$($windowsVersionMapping[$wv])") > $null
-    $tags.Add("sitecore-redis:3.0.504-windowsservercore-$($wv)") > $null
+    $defaultTags | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
+
+    if ($Topology -contains "xp" -or $Topology -contains "xc")
+    {
+        $miscTags | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
+    }
 
     foreach ($scv in $SitecoreVersion)
     {
-        $tags.Add("sitecore-assets:$($scv)-nanoserver-$($windowsVersionMapping[$wv])") > $null
-        if ([Version]$scv -ge [Version]"9.3.0" -and [int]$windowsVersionMapping[$wv] -gt 1809)
-        {
-            $tags.Add("sitecore-ps:$($scv)-nanoserver-$($windowsVersionMapping[$wv])") > $null
-        }
+        $assetTags | SitecoreFilter -Version $scv | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
 
-        if ($Topology -contains "xm")
+        if($Topology -contains "xm")
         {
-            $tags.Add("sitecore-xm-cd:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xm-cm:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xm-solr:$($scv)-nanoserver-$($windowsVersionMapping[$wv])") > $null
-            $tags.Add("sitecore-xm-sqldev:$($scv)-windowsservercore-$($wv)") > $null
+            $xmTags | SitecoreFilter -Version $scv | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
         }
 
-        if ($Topology -contains "xp")
+        if($Topology -contains "xp")
         {
-            $tags.Add("sitecore-xp-cd:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-identity:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-solr:$($scv)-nanoserver-$($windowsVersionMapping[$wv])") > $null
-            $tags.Add("sitecore-xp-standalone:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-sqldev:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-xconnect:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-xconnect-automationengine:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-xconnect-indexworker:$($scv)-windowsservercore-$($wv)") > $null
-            $tags.Add("sitecore-xp-xconnect-processingengine:$($scv)-windowsservercore-$($wv)") > $null
+            $xpTags | SitecoreFilter -Version $scv | WindowsFilter -Version $wv | ForEach-Object { $tags.Add($_) > $null }
         }
 
-        if ($Topology -contains "xc")
+        if($Topology -contains "xc")
```

Comment:
Could all of these if ($Topology) -> $Tags.Add be refactored to a function "TagAdder" :)

Seems like this is getting a little complicated

Reply:
I'm not sure how to make this much cleaner at the moment.